capsules/extra/src/ieee802154/driver.rs

```diff
@@ -1037,17 +1096,68 @@ fn encode_address(addr: &Option<MacAddress>) -> usize {
             let read_present = kernel_data
                 .get_readwrite_processbuffer(rw_allow::READ)
                 .and_then(|read| {
                     read.mut_enter(|rbuf| {
-                        let len = min(rbuf.len(), data_offset + data_len);
-                        // Copy the entire frame over to userland, preceded by two
-                        // bytes: the data offset and the data length.
-                        rbuf[..len].copy_from_slice(&buf[..len]);
-                        rbuf[0].set(data_offset as u8);
-                        rbuf[1].set(data_len as u8);
+                        ///////////////////////////////////////////////////////////////////////////////////////////
+                        // NOTE: context for the ring buffer and assumptions regarding the ring buffer
+                        // format and usage can be found in the detailed comment at the top of this file.
+                        //      Ring buffer format:
+                        //          | read index | write index | user_frame 0 | user_frame 1 | ... | user_frame n |
+                        //      user_frame format:
+                        //          | data_offset | data_len | mic_len | 15.4 frame |
+                        ///////////////////////////////////////////////////////////////////////////////////////////
+
+                        // 2 bytes for the readwrite buffer metadata (read / write index)
+                        const RING_BUF_METADATA_SIZE: usize = 2;
+
+                        // Confirm the availability of the buffer. A buffer of len 0 is indicative
+                        // of the userprocess not allocating a readwrite buffer. We must also
+                        // confirm that the userprocess correctly formatted the buffer to be of length
+                        // 2 + n * USER_FRAME_MAX_SIZE, where n is the number of user frames that the
+                        // buffer can store.
+                        if rbuf.len() == 0
+                            || (rbuf.len() - RING_BUF_METADATA_SIZE) % USER_FRAME_MAX_SIZE != 0
+                        {
+                            // kernel::debug!("[15.4 Driver] Error - improperly formatted readwrite buffer provided");
+                            return false;
+                        }
+
+                        let mic_len = header.security.map_or(0, |sec| sec.level.mic_len());
+                        let frame_len = data_offset + data_len + mic_len;
+
+                        // Frame length incorporates the PSDU offset, so we must subtract this value
+                        // in addition to incorporating the user frame metadata size to obtain the
+                        // length of the user frame.
+                        let user_frame_len = frame_len + USER_FRAME_METADATA_SIZE - PSDU_OFFSET;
+
+                        let read_index = rbuf[0].get() as usize;
+                        let mut write_index = rbuf[1].get() as usize;
```

Comment:
This is concerning. While it's maybe ok to store metadata in a userspace buffer, and tell userspace to not mess with it, using the `[]` operator on slices means it's trivial for userspace to crash the kernel by changing this byte.

One fix is to not use `[]` and instead use get/set. A less rusty version would be to check the indices.

Reply:
Good catch, I overlooked this. I'll go ahead and add this.